sky/provision/kubernetes/instance.py

```diff
@@ -0,0 +1,713 @@
+                'from the image.')
+
+
+def _setup_ssh_in_pods(namespace: str, new_nodes: List) -> None:
+    # Setting up ssh for the pod instance. This is already setup for
+    # the jump pod so it does not need to be run for it.
+    set_k8s_ssh_cmd = [
+        '/bin/sh',
+        '-c',
+        (
+            'prefix_cmd() '
+            '{ if [ $(id -u) -ne 0 ]; then echo "sudo"; else echo ""; fi; }; '
+            'export DEBIAN_FRONTEND=noninteractive;'
+            '$(prefix_cmd) apt-get update;'
+            '$(prefix_cmd) apt install openssh-server rsync -y; '
+            '$(prefix_cmd) mkdir -p /var/run/sshd; '
+            '$(prefix_cmd) '
+            'sed -i "s/PermitRootLogin prohibit-password/PermitRootLogin yes/" '
+            '/etc/ssh/sshd_config; '
+            '$(prefix_cmd) sed '
+            '"s@session\\s*required\\s*pam_loginuid.so@session optional '
+            'pam_loginuid.so@g" -i /etc/pam.d/sshd; '
+            'cd /etc/ssh/ && $(prefix_cmd) ssh-keygen -A; '
+            '$(prefix_cmd) mkdir -p ~/.ssh; '
+            '$(prefix_cmd) cp /etc/secret-volume/ssh-publickey '
+            '~/.ssh/authorized_keys; '
+            '$(prefix_cmd) chown -R $(whoami) ~/.ssh;'
+            '$(prefix_cmd) chmod 700 ~/.ssh; '
+            '$(prefix_cmd) chmod 600 ~/.ssh/authorized_keys; '
+            '$(prefix_cmd) service ssh restart; '
+            # Eliminate the error
+            # `mesg: ttyname failed: inappropriate ioctl for device`.
+            # See https://www.educative.io/answers/error-mesg-ttyname-failed-inappropriate-ioctl-for-device  # pylint: disable=line-too-long
+            '$(prefix_cmd) sed -i "s/mesg n/tty -s \\&\\& mesg n/" ~/.profile;')
+    ]
+    # TODO(romilb): We need logging and surface errors here.
+    for new_node in new_nodes:
+        run_command_on_pods(new_node.metadata.name, namespace, set_k8s_ssh_cmd)
+
+
+def _label_pod(namespace: str, pod_name: str, label: Dict[str, str]) -> None:
+    """Label a pod."""
+    kubernetes.core_api().patch_namespaced_pod(
+        pod_name,
+        namespace, {'metadata': {
+            'labels': label
+        }},
+        _request_timeout=kubernetes.API_TIMEOUT)
+
+
+def _setup_service(namespace: str, service_spec: Optional[Dict[str, Any]],
+                   pods: List) -> None:
+    if service_spec is not None:
+        logger.info('run_instances: calling create_namespaced_service '
+                    f'(count={len(pods)}).')
+
+        for pod in pods:
+            node_uuid = pod.meta.labels.get(TAG_SVC_POD_UUID)
+            metadata = service_spec.get('metadata', {})
+            metadata['name'] = pod.meta.name
+            service_spec['metadata'] = metadata
+            service_spec['spec']['selector'] = {TAG_SVC_POD_UUID: node_uuid}
+            _ = kubernetes.core_api().create_namespaced_service(
+                namespace, service_spec)
+            # TODO(zhwu): Should we wait for the service to be ready?
```

Comment:
We probably don't need to wait since service creation is nearly instant, only the endpoint mapping to a real pod takes some time (which is fine, even if it takes too long (which is unlikely) our SSH retry will account for this).